components/net/image_cache.rs

```diff
@@ -433,33 +488,121 @@ struct ImageCacheStore {
+    /// images that do not have a key yet.
+    key_cache: KeyCache,
 }
 
 impl ImageCacheStore {
-    // Change state of a url from pending -> loaded.
-    fn complete_load(&mut self, key: LoadKey, mut load_result: LoadResult) {
+    /// Finishes loading the image by setting the WebRenderImageKey and calling `compete_load` or `complete_load_svg`.
+    fn set_key_and_finish_load(&mut self, pending_image: PendingKey, image_key: WebRenderImageKey) {
+        match pending_image {
+            PendingKey::RasterImage((pending_id, mut raster_image)) => {
+                set_webrender_image_key(&self.compositor_api, &mut raster_image, image_key);
+                self.complete_load(pending_id, LoadResult::LoadedRasterImage(raster_image));
+            },
+            PendingKey::Svg((pending_id, mut raster_image, requested_size)) => {
+                set_webrender_image_key(&self.compositor_api, &mut raster_image, image_key);
+                self.complete_load_svg(raster_image, pending_id, requested_size);
+            },
+        }
+    }
+
+    /// If a key is available the image will be immediately loaded, otherwise it will load then the next batch of
+    /// keys is received. Only call this if the image does not have a `LoadKey` yet.
+    fn load_image_with_keycache(&mut self, pending_image: PendingKey) {
+        if let Some(pipeline_id) = self.pipeline_id {
+            match self.key_cache.cache {
+                KeyCacheState::PendingBatch => {
+                    self.key_cache.images_pending_keys.push_back(pending_image);
+                },
+                KeyCacheState::Ready(ref mut cache) => match cache.pop() {
+                    Some(image_key) => {
+                        self.set_key_and_finish_load(pending_image, image_key);
+                    },
+                    None => {
+                        self.key_cache.images_pending_keys.push_back(pending_image);
+                        self.compositor_api.generate_image_key_async(pipeline_id);
+                        self.key_cache.cache = KeyCacheState::PendingBatch
+                    },
+                },
+            }
+        } else {
+            error!("No pipeline id for this image key cache. Have to block.");
+        }
+    }
+
+    /// Insert received keys into the cache and complete the loading of images.
+    fn insert_keys_and_load_images(&mut self, image_keys: Vec<WebRenderImageKey>) {
+        if let KeyCacheState::PendingBatch = self.key_cache.cache {
+            let len = min(
+                self.key_cache.cache.size(),
+                self.key_cache.images_pending_keys.len(),
+            );
+            self.key_cache.cache = KeyCacheState::Ready(image_keys);
+            let images = self
+                .key_cache
+                .images_pending_keys
+                .drain(0..len)
+                .collect::<Vec<PendingKey>>();
+            for key in images {
+                self.load_image_with_keycache(key);
+            }
```

Comment:
Here you forgot to address https://github.com/servo/servo/pull/37369/files#r2151670315